routing/routing_helpers.cpp

```diff
@@ -18,6 +20,84 @@ namespace routing
 using namespace std;
 using namespace traffic;
 
+void FillSegmentInfo(vector<Segment> const & segments, vector<Junction> const & junctions,
+                     Route::TTurns const & turns, Route::TStreets const & streets,
+                     Route::TTimes const & times, shared_ptr<TrafficStash> const & trafficStash,
+                     vector<RouteSegment> & routeSegment)
+{
+  CHECK_EQUAL(segments.size() + 1, junctions.size(), ());
+  CHECK(!turns.empty(), ());
+  CHECK(!times.empty(), ());
+
+  CHECK(std::is_sorted(turns.cbegin(), turns.cend(), my::LessBy(&turns::TurnItem::m_index)), ());
+  CHECK(std::is_sorted(streets.cbegin(), streets.cend(), my::LessBy(&Route::TStreetItem::first)), ());
+  CHECK(std::is_sorted(times.cbegin(), times.cend(), my::LessBy(&Route::TTimeItem::first)), ());
+
+  CHECK_LESS(turns.back().m_index, junctions.size(), ());
+  if (!streets.empty())
+    CHECK_LESS(streets.back().first, junctions.size(), ());
+  CHECK_LESS(times.back().first, junctions.size(), ());
+
+  routeSegment.clear();
+  if (segments.empty())
+    return;
+
+  routeSegment.reserve(segments.size());
+  // Note. |turns|, |streets| and |times| have point based index.
+  // On the other hand turns, street names and times are considered for further end of |segments| after conversion
+  // to |segmentInfo|. It means that street, turn and time information of zero point will be lost after
+  // conversion to |segmentInfo|.
+  size_t turnIdx = turns[0].m_index != 0 ? 0 : 1;
```

Comment:
The reason is |turns[0]| contains information about the first turn of the route. Generally speaking the first turn of the route may lie on any point of the route including point number 0 (if so turns[0].m_index == 0).

If we want to skip the first point of the route we need to skip turns[0] if there's a turn at the route start (turns[0].m_index == 0) and we need to leave turns[0] for any other cases (turns[0].m_index > 0).

Note. The size of |turns| is equal to number of turns at the route + 1 (ReachedYourDestination). It could be much less then number of route points.